libff/algebra/curves/bls12_381/bls12_381_g2.hpp

```diff
@@ -0,0 +1,130 @@
+/** @file
+ *****************************************************************************
+ * @author     This file is part of libff, developed by SCIPR Lab
+ *             and contributors (see AUTHORS).
+ * @copyright  MIT license (see LICENSE file)
+ *****************************************************************************/
+
+#ifndef BLS12_381_G2_HPP_
+#define BLS12_381_G2_HPP_
+#include <libff/algebra/curves/bls12_381/bls12_381_init.hpp>
+#include <libff/algebra/curves/curve_utils.hpp>
+#include <vector>
+
+namespace libff
+{
+
+class bls12_381_G2;
+std::ostream &operator<<(std::ostream &, const bls12_381_G2 &);
+std::istream &operator>>(std::istream &, bls12_381_G2 &);
+
+class bls12_381_G2
+{
+public:
+#ifdef PROFILE_OP_COUNTS
+    static long long add_cnt;
+    static long long dbl_cnt;
+#endif
+    static std::vector<std::size_t> wnaf_window_table;
+    static std::vector<std::size_t> fixed_base_exp_window_table;
+    static bls12_381_G2 G2_zero;
+    static bls12_381_G2 G2_one;
+    static bls12_381_Fq2 coeff_a; // VV
+    static bls12_381_Fq2 coeff_b; // VV
+
+    // Cofactor
+    static const mp_size_t h_bitcount = 507;
+    static const mp_size_t h_limbs =
+        (h_bitcount + GMP_NUMB_BITS - 1) / GMP_NUMB_BITS;
+    static bigint<h_limbs> h;
+
+    typedef bls12_381_Fq base_field;
+    typedef bls12_381_Fq2 twist_field;
+    typedef bls12_381_Fr scalar_field;
+
+    bls12_381_Fq2 X, Y, Z;
+
+    // using Jacobian coordinates
+    bls12_381_G2();
+    bls12_381_G2(
+        const bls12_381_Fq2 &X, const bls12_381_Fq2 &Y, const bls12_381_Fq2 &Z)
+        : X(X), Y(Y), Z(Z){};
+
+    static bls12_381_Fq2 mul_by_b(const bls12_381_Fq2 &elt);
+
+    void print() const;
+    void print_coordinates() const;
+
+    void to_affine_coordinates();
+    void to_special();
+    bool is_special() const;
+
+    bool is_zero() const;
+
+    bool operator==(const bls12_381_G2 &other) const;
+    bool operator!=(const bls12_381_G2 &other) const;
+
+    bls12_381_G2 operator+(const bls12_381_G2 &other) const;
+    bls12_381_G2 operator-() const;
+    bls12_381_G2 operator-(const bls12_381_G2 &other) const;
+
+    bls12_381_G2 add(const bls12_381_G2 &other) const;
+    bls12_381_G2 mixed_add(const bls12_381_G2 &other) const;
+    bls12_381_G2 dbl() const;
+    bls12_381_G2 mul_by_q() const;
+    bls12_381_G2 untwist_frobenius_twist() const; // from bls12_377 (VV)
```

Comment:
I think this is not required for bls12-381 since we use the generic subgroup checks.